gridGeneration.py

```diff
@@ -0,0 +1,259 @@
+                                np_array[coordinate_ind, :][1] >= in_grids[ind][1] and np_array[coordinate_ind, :][1] < in_grids[ind][3]):
+                                tmp_array[ind] += 1
+                
+                #print('temp :', tmp_array)
+                for ind2 in range(len(in_grids)):
+                    if tmp_array[ind2] > 0:
+                        road_counts_his[ind2] += 1
+        # ==========================================
+        # write out a csv file
+        csv_matrix = []
+        csv_matrix.append(['grid_id', 'err_roads', 'road_counts'])
+        for index in range(len(in_grids)):
+            csv_matrix.append([index + 1, in_counts[index], road_counts_his[index]])
+            
+        with open(os.path.join(out_folder, 'road-' + os.path.basename(out_folder) + '.csv'), "w") as out_f:
+            writer = csv.writer(out_f)
+            writer.writerows(csv_matrix)
+# =======================================
+# Compute cell size
+def cell_size_computation(depth_val, bounding_box_collec):
+    print('Level ', depth_val)
+    print('Grid size ( Width (X length) x Height (Y length) ): ', (bounding_box_collec[0][2] - bounding_box_collec[0][0]
+          , bounding_box_collec[0][3] - bounding_box_collec[0][1]))
+    print('========================')
+# =======================================
+def main():
+    # declare arguments and variables
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--maxDepth', type = str, default='', help='max depth of a k-d tree')
+    parser.add_argument('--countNum', type = str, default='', help='a count value')
+    parser.add_argument('--gridPercent', type = str, default='', help='a grid percentage')
+    parser.add_argument('--maxCount', type = str, default='', help='maximum count to the second k-d tree')
+    parser.add_argument('--outFolder', type = str, default='', help='path to an ouput folder')
+
+    args = parser.parse_args()
+    maximum_level = args.maxDepth
+    folder_path = args.outFolder
+    count_num = int(args.countNum)
+    grid_percent = float(args.gridPercent)
+
+    print(os.path.join(folder_path, 'histogram'))
+    
+    path = 'histogram'
+    geojson_path = 'geojson'
+    
+    if not os.path.exists(os.path.join(folder_path, path)):
+        print('Create the histogram directory !!')
+        os.makedirs(os.path.join(folder_path, path))
+    if not os.path.exists(os.path.join(folder_path, geojson_path)):
+        print('Create the geojson directory !!')
+        os.makedirs(os.path.join(folder_path, geojson_path))
+    
+    max_count = 0
+    flag_val = False
+    input_data = None
+    optimal_count_list = None
+    optimal_grid_size_list = None
+    first_depth = 0
+    big_initial_area = 0.0
+
+    if args.maxCount:
+        max_count = int(args.maxCount)
+
+    # read L-2 dissimilarity from stand input
+    for line in sys.stdin.readlines():
+        input_data = ast.literal_eval(line)
```

Comment:
It seems a bit odd to be printing a huge amount of info to stdout, only to then ingest it into another script. It would be better if this script called a function from geoProcess and got the data as a return, or the data was saved to a file (or files) by geoProcess and this gathered it from there (especially since some of the data is just being immediately written to a file anyway).

Reply:
remove the shell script, and treat geoProcess.py as a module. The main program is essentially able to call a function in the geoProcess module and return values.